### PR TITLE
bpo-30766: Make CHECK_STATUS_PTHREAD signal-safe.

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

File: Python/thread_pthread.h [hunk 1]
```diff
@@ -142,9 +142,19 @@ typedef struct {
     pthread_mutex_t  mut;
 } pthread_lock;
 
+#define PUTS(fd, str) _Py_write_noraise(fd, str, (int)strlen(str))
```
Comment (vstinner) [inline Python/thread_pthread.h:145]: Why casting strlen() to int?

File: Python/thread_pthread.h [hunk 2]
```diff
@@ -142,9 +142,19 @@ typedef struct {
...
+
 #define CHECK_STATUS(name)  if (status != 0) { perror(name); error = 1; }
-#define CHECK_STATUS_PTHREAD(name)  if (status != 0) { fprintf(stderr, \
-    "%s: %s\n", name, strerror(status)); error = 1; }
+/* CHECK_STATUS_PTHREAD is async-signal-safe */
+#define CHECK_STATUS_PTHREAD(name) \
```
Comment (vstinner) [inline Python/thread_pthread.h:149]: I never understood the purpose of these functions. Do we really need these checks? Why not replacing them with an assertion?

If you want to get them at runtime, maybe call Py_FatalError()?

File: Python/thread_pthread.h (hunk 1 above)
Comment (pitrou) [reply Python/thread_pthread.h:145]: I don't know, I copied this macro from `traceback.c`.

File: Python/thread_pthread.h (hunk 2 above)
Comment (pitrou) [reply Python/thread_pthread.h:149]: I don't know. I'm inclined not to remove them right now.
OTOH, this whole PR may be made obsolete by https://github.com/python/cpython/pull/2408.

Comment (vstinner) [reply Python/thread_pthread.h:149]: > I don't know. I'm inclined not to remove them right now. OTOH, this whole PR may be made obsolete by #2408.

Oh no, please finish your work. I like the idea of more reliable locks :-)

**Comment by pitrou** [comment]
I'm inclined to reject this PR. Locks are not signal-safe, so it doesn't make sense to fix a tiny bit of code inside them.

**Pull request closed without merging**